### ISSUE TITLE
[CLOSED] Introduce GC operation

### ISSUE THREAD

**Issue by aionbot** [aionnetwork/AVM#234]
**Issue created by [yulongaion](https://github.com/yulongaion)** (on Tuesday Sep 04, 2018 at 20:24 GMT)

We need to add deterministic and constrained GC operation, with a proper economic model.

**Comment by aionbot** [comment]
**Comment by [jeff-aion](https://github.com/jeff-aion)** (on Wednesday Sep 12, 2018 at 19:49 GMT)

Note that this depends on #246:

- if we proceed with an object graph DB approach, this GC will be implemented at the storage-level
- if we continue to back-end onto the key-value store, a GC would have to be implemented at the VM-level
- if we want to introduce a proper economic model, the concept of a GC needs to be blockchain-level

While the GC being part of the storage system allows for a better division of concerns and increases the freedom of implementation of the GC, if we make it fully-transparent, we lose the ability to make an economic incentive for the application to avoid memory leaks.  Even VM-level implementations which intend to be transparent, have this problem.

This means that we either need to make the GC operation a blockchain-level idea (even though we could still implement it at the storage level - so long as the result could be described in a spec), or defer this consideration to see if it becomes a problem in the future (which is essentially what Ethereum did).

**Comment by aionbot** [comment]
**Comment by [jeff-aion](https://github.com/jeff-aion)** (on Wednesday Sep 26, 2018 at 14:53 GMT)

Prototyping this GC has been generally not too complicated but it has raised an interesting question:  who is responsible for creating and maintaining the storage abstraction reference?  This is specifically of interest within the context of the hot DApp cache:  if we recreate the storage abstraction for each call, but also keep it in the `LoadedDApp` in the cache, then it is possible that we can find conflicting versions of the storage:  one which has GC-ed, and one which has not.

I suspect that the storage abstraction reference should be _owned_ by the `LoadedDApp` since we will need to check something out of the storage cache in order to GC it, anyway.

**Comment by aionbot** [comment]
**Comment by [jeff-aion](https://github.com/jeff-aion)** (on Wednesday Sep 26, 2018 at 15:58 GMT)

The basic GC can be seen here:  12325004a43e0adac1be6543c10819452e6c21cd

This GC acts as a copying semi-space collector (a "scavenger"), using a breadth-first scanning order.  This is relatively easy to define, meaning consensus can be maintained if we want to use the allocation-order-defined Merkle tree for storage root hash (since such a collector re-orders objects, we need to define the ordering scheme).

If we were to use something like the streaming delta hash, the order wouldn't matter.

One of the other advantages of this approach is limiting the amount of in-memory data which must be tracked.  Since objects are reloaded to be scanned, the only in-memory structure we keep is the old-to-new reference fixup map (which could be stored in the original objects, after moving, if we were worried about this map size but it is bounded by the number of objects in the live-set).

**Comment by aionbot** [comment]
**Comment by [jeff-aion](https://github.com/jeff-aion)** (on Wednesday Sep 26, 2018 at 17:13 GMT)

The prototyping work here (regarding the GC implementation) will be broken down into smaller commits and merged into master:  https://github.com/aionnetworkp/aion_vm/tree/storage_prototyping
